File: hash/Solution15.java
import java.util.List;
import java.util.LinkedList;
import java.util.Arrays;

class Solution15 {
    public List<List<Integer>> threeSum(int[] nums) {
    	List<List<Integer>> result = new LinkedList<>();
    	Arrays.sort(nums);
    	for(int i = 0; i < nums.length; i++){
    		if(i != 0 && nums[i] == nums[i - 1]){
    			continue;
    		}
            int target = -nums[i];
    		for(int j = i + 1; j < nums.length - 1; j++){
    			if(j != i + 1 && nums[j - 1] == nums[j]){
    				continue;
    			}
    			int point = nums.length - 1;
    			while(nums[point] + nums[j] > target && point > j + 1){
    				point -= 1;
    			}
    			if(nums[j] + nums[point] == target){
    				List<Integer> cur = new LinkedList<>();
    				cur.add(nums[i]);
    				cur.add(nums[j]);
    				cur.add(nums[point]);
    				result.add(cur);
    			}
    		}
    	}
    	return result;
    }

    public static void main(String[] args){
    	Solution15 s = new Solution15();
    	int[] test = new int[]{-1,0,1,2,-1,-4};
    	List<List<Integer>> result = s.threeSum(test);
    }
}

File: C++/Solution15.cpp
#include<iostream>
#include<vector>

#include<algorithm>

using namespace std;

class Solution15 {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int size = nums.size();
        vector<vector<int>> result;
        sort(nums.begin(), nums.end());
        for(int index1 = 0; index1 < size; index1++){
            if(index1 > 0 && nums[index1] == nums[index1 - 1]){
                continue;
            }
            if(nums[index1] > 0){
                break;
            }
            int target = -nums[index1];
            int index2 = index1 + 1, index3 = size - 1;
            while(index3 > index2){
                if(nums[index2] + nums[index3] == target){
                    result.push_back(vector<int>{nums[index1], nums[index2], nums[index3]});
                    index2++;
                    index3--;
                    while(index2 < size && nums[index2] == nums[index2 - 1]){
                        index2++;
                    }
                    while(index3 >= 0 && nums[index3] == nums[index3 + 1]){
                        index3--;
                    }
                }
                else if(nums[index2] + nums[index3] > target){
                    index3--;
                    while(index3 >= 0 && nums[index3] == nums[index3 + 1]){
                        index3--;
                    }
                }
                else{
                    index2++;
                    while(index2 < size && nums[index2] == nums[index2 - 1]){
                        index2++;
                    }
                }
            }
        }
        return result;
    }
};

int main(){
    Solution15 s;
    vector<int> test = {1};
    vector<vector<int>> result = s.threeSum(test);
    return 0;
}